submission/races.go

```diff
@@ -0,0 +1,284 @@
+	// PostBatchInterval is duration between parallel batch call and subsequent requests to Logs within group.
+	// TODO(Mercurrent): optimize to avoid excessive requests.
+	PostBatchInterval = time.Second
+)
+
+// submissionResult holds outcome of a single-log submission.
+type submissionResult struct {
+	sct *ct.SignedCertificateTimestamp
+	err error
+}
+
+type groupState struct {
+	Name    string
+	Success bool
+}
+
+// safeSubmissionState is a submission state-machine for set of Log-groups. When some group is complete cancels all requests that are not needed by any group.
+type safeSubmissionState struct {
+	logToGroups map[string]map[string]bool
+	groupNeeds  map[string]int
+
+	results map[string]*submissionResult
+	cancels map[string]context.CancelFunc
+	mu      sync.Mutex
+}
+
+func newSafeSubmissionState(groups ctpolicy.LogPolicyData) *safeSubmissionState {
+	var s safeSubmissionState
+	s.logToGroups = ctpolicy.GroupByLogs(groups)
+	s.groupNeeds = make(map[string]int)
+	for _, g := range groups {
+		s.groupNeeds[g.Name] = g.MinInclusions
+	}
+	s.results = make(map[string]*submissionResult)
+	s.cancels = make(map[string]context.CancelFunc)
+	return &s
+}
+
+// request includes empty submissionResult in the set, returns whether the entry wasn't requested (added) before.
+func (sub *safeSubmissionState) request(logURL string, cancel context.CancelFunc) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sub.results[logURL] != nil {
+		// Already requested.
+		return false
+	}
+	sub.results[logURL] = &submissionResult{}
+	isAwaited := false
+	for g := range sub.logToGroups[logURL] {
+		if sub.groupNeeds[g] > 0 {
+			isAwaited = true
+			break
+		}
+	}
+	if !isAwaited {
+		// No groups expecting result from this Log.
+		return false
+	}
+	sub.cancels[logURL] = cancel
+	return true
+}
+
+// returns whether the Log has already been requested.
+func (sub *safeSubmissionState) has(logURL string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	return sub.results[logURL] != nil
+}
+
+// Gets SCT-result. Writes it down if it is error or awaited-SCT. Re-calculates group-completion and cancels any running but not-awited-anymore Log-requests.
+func (sub *safeSubmissionState) setResult(logURL string, sct *ct.SignedCertificateTimestamp, err error) {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sct == nil {
+		sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		return
+	}
+	// If at least one group needs that SCT, result is set. Otherwise dumped.
+	for groupName := range sub.logToGroups[logURL] {
+		if sub.groupNeeds[groupName] > 0 {
+			sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		}
+		sub.groupNeeds[groupName]--
+	}
+
+	// Cancel any pending Log-requests for which there's no more awaiting Log-group.
+	for logURL, groupSet := range sub.logToGroups {
+		isAwaited := false
+		for g := range groupSet {
+			if sub.groupNeeds[g] > 0 {
+				isAwaited = true
+				break
+			}
+		}
+		if !isAwaited && sub.cancels[logURL] != nil {
```

Comment:
It's not wrong to use nil in maps like this but I think it's less common in Go than others.

Reply:
Do you have any suggestions?